go-controller/pkg/ovn/controller/admin_network_policy/repair.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+	anpapi "sigs.k8s.io/network-policy-api/apis/v1alpha1"
+)
+
+// repairAdminNetworkPolicies is called at startup and as the name suggests
+// aims to repair the NBDB logical objects -> port groups, acls and address-sets
+// that are created for the admin network policies in the cluster
+// Logic:
+// We fetch all the ANPs present in the cluster from the Lister
+// We fetch PGs, ACLs, AddressSets that are owned by ANP objectIDs based on
+// externalIDs match from the NBDB. Using predicate search we check if
+// the relevant ANP still exists for these objects and if not we delete them
+// TODO (tssurya): Handle cleaning up stale IPs in the address-sets by fetching all the relevant
+// peer pods that are being served by ANPs - pods could stop matching when ovnkube-controller is down
+// TODO2 (tssurya): Handle cleaning up stale ports from PGs by fetching all the relevant
+// subject pods that are being served by ANPs - pods could stop matching when ovnkube-controller is down
+func (c *Controller) repairAdminNetworkPolicies() error {
+	start := time.Now()
+	defer func() {
+		klog.Infof("Repairing admin network policies took %v", time.Since(start))
+	}()
+	c.Lock()
+	defer c.Unlock()
+	anps, err := c.anpLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list ANPs from the lister, err: %v", err)
+	}
+	existingANPs := map[string]*anpapi.AdminNetworkPolicy{}
+	for _, anp := range anps {
+		existingANPs[anp.Name] = anp
+	}
+
+	// Deal with PortGroup Repairs first
+	// We grab all the port groups that belong to ANP controller using externalIDs
+	// and compare the value with the name of existing ANPs. If no match is found
+	// we delete that port group along with all the acls in it.
+	p := func(pg *nbdb.PortGroup) bool {
+		anpName, ok := pg.ExternalIDs[ANPExternalIDKey]
+		if !ok {
+			return false // we don't care about this PG as it doesn't belong to ANP controller
+		}
+		_, ok = existingANPs[anpName]
+		return !ok // return if it doesn't exist in the cache
+	}
+	stalePGs, err := libovsdbops.FindPortGroupsWithPredicate(c.nbClient, p)
+	if err != nil {
+		return fmt.Errorf("unable to fetch port groups by predicate, err: %v", err)
+	}
+	if len(stalePGs) > 0 {
+		klog.Infof("Deleting Stale PortGroups +%v", stalePGs)
+		err = libovsdbops.DeletePortGroupsWithPredicate(c.nbClient, p)
+		if err != nil {
+			return fmt.Errorf("unable to delete stale port groups, err: %v", err)
+		}
+	}
+
+	// Deal with ACL Repairs
+	// We grab all the ACLs that belong to ANP controller using externalIDs
+	// and compare with the rules we have in existing ANPs. The ones that don't match
+	// will be deleted from the corresponding PGs.
+	// NOTE: When we call syncAdminNetworkPolicy function after this for every ANP on startup,
+	// the right ACLs will be recreated.
```

Comment:
The controller itself would be able to do this on the Add events. I guess there is no problem in doing it here but it looks like a missed opportunity to share code given that we are able to do it for Update events that remove rules.
The same for AS.
So it looks like the repair could potentially deal only with deleting PGs of deleted policies.

Reply:
yea I see your point. as of now yeah I am accounting only for cleanup logic in repair.
I could invoke an add for every ANP on repair like you said  - the functions already exist for which specific ACLs and AS-es we want to fix..., Let me see if I can do this on next update; leaving comment unresolved because this has not been addressed yet.

Reply:
so i was talking to Patryk and he pointed out something important - the whole reason behind us having a repair on startup is to cleanup things that are stale in DB and don't exist in the KAPI.
So if I get a add ANP, OVNK KILLED/DOWN, DELETE ANP, OVNK HEALS and startup again - I won't get the event back for that deleted object - so update code in ANP controller cannot take care of that since that key will never come back agian... I think we should keep the repair cleanup.

Reply:
If I got that right, we only need to cleanup 
1. stale policies, because we won't get any events for them
2. stale parts of existing policies, that won't be cleaned up via add event handling with empty cache (there is an alternative approach to build cache based on the db state, but we are not doing that)

First case I think is simple and we all agree on, just delete port group and address sets (acls will be garbage collected).
For the second part, I checked `createNewANP`, and it looks like it will take care of updating port group, acls, and address sets that still exist. The only thing that is missing is deleting stale address sets, that are not needed anymore (as in `constructOpsForRuleChanges`). I don't see that case being handled in repair yet, but I think these are all the resources that need cleanup.

What do you guys think?

Reply:
> If I got that right, we only need to cleanup
> 
> 1. stale policies, because we won't get any events for them
> 2. stale parts of existing policies, that won't be cleaned up via add event handling with empty cache (there is an alternative approach to build cache based on the db state, but we are not doing that)

yes exactly and the reason we don't do the alternative approach to build cache based on db state is because we want to build cache based on kapi state. We should treat kapi as source of truth and cleanup db state and not the other way around.

> 
> First case I think is simple and we all agree on, just delete port group and address sets (acls will be garbage collected). 

yes I agree with this and this is being done in repair already.

> For the second part, I checked `createNewANP`, and it looks like it will take care of updating port group, acls, and address sets that still exist. 

yeah so I think currently after the previous review rounds, we call `createNewANP` which is re-setting the ACLs totally -> `pg := libovsdbops.BuildPortGroup(portGroupName, desiredPorts, desiredACLs, pgExternalIDs)` where `BuildPortGroup`  is building the PG from scratch for every add thus wiping out stale ACLs. That means I could just remove the parts of repair ACLs code since there won't be `ruleChanges` stuff that are not handled by createANP..

The only thing that is missing is deleting stale address sets, that are not needed anymore (as in `constructOpsForRuleChanges`). I don't see that case being handled in repair yet, but I think these are all the resources that need cleanup.
> 
> What do you guys think?

+1 yeah and I can re-use that code in the repair to clean up address-sets for existing policies...BUT we cannot do this unless we also cleanup those specific ACLs in the repair before the address-sets for avoiding the acl/address-set integrity violation. I feel this cleanup should be done in the repair and not in `createNewANP` (because fresh add and add after restart should all be dealt like fresh add...) but i am ok with whatever/whereever you both feel this should be done at.

Reply:
yeah to avoid integrity violation it makes sense to delete stale acls in repair too (and maybe a comment about this reasoning so we don't try to delete later)

>yes exactly and the reason we don't do the alternative approach to build cache based on db state is because we want to build cache based on kapi state. We should treat kapi as source of truth and cleanup db state and not the other way around.

I mean we can build cache as the existing state first without any other repair, and then on add events that we receive on informer start, we handle add events as updates that will include cleanup (similar to the current update event handling)

Reply:
> yeah to avoid integrity violation it makes sense to delete stale acls in repair too (and maybe a comment about this reasoning so we don't try to delete later)
> 
> > yes exactly and the reason we don't do the alternative approach to build cache based on db state is because we want to build cache based on kapi state. We should treat kapi as source of truth and cleanup db state and not the other way around.
> 
> I mean we can build cache as the existing state first without any other repair, and then on add events that we receive on informer start, we handle add events as updates that will include cleanup (similar to the current update event handling)

oh i see, so we will initialize the cache with existing DB stuff and then since its already pre-populated when we get the add events its going to be just treating it as updates and our updates are taking care of all scenarios.. so if we do this, then i can remove the ACL and Address-sets repair for "rule changes" totally. This will/should work except my caches are not storing the libovsdb cache state in currentState; its storing the previous memory state of the kapi, so if i do this now curent=desired on startup for existing policies.. let me for now do this removal in the repair explicitly and leave a comment on why we do it this way.

When I redo the caches we can also change the way we do this on startup

Reply:
So what I meant with my comment is that it looks like the create ANP code path should be able to clean up stale AS just as the update ANP code path does, and that might be an opportunity to share more code between them.

At the same time the repair could only deal with stale entries from deleted ANPs only, which is a simpler matter.

Now it just looks like we have the complication of cleaning up stale AS for existing policies in two places, the repair and the controller ANP update code path. 

> yes exactly and the reason we don't do the alternative approach to build cache based on db state is because we want to build cache based on kapi state. We should treat kapi as source of truth and cleanup db state and not the other way around.

Please note that when I talk about the DB state option, I don't mean it to build a (new) cache out of it or for it to be the source of truth. It is just a mean to compare desired state to current state and apply a diff.

Reply:
> oh i see, so we will initialize the cache with existing DB stuff and then since its already pre-populated when we get the add events its going to be just treating it as updates and our updates are taking care of all scenarios.. so if we do this, then i can remove the ACL and Address-sets repair for "rule changes" totally. This will/should work except my caches are not storing the libovsdb cache state in currentState; its storing the previous memory state of the kapi, so if i do this now curent=desired on startup for existing policies.. let me for now do this removal in the repair explicitly and leave a comment on why we do it this way.

You don't need that. This is the code that does it for updated ANPs
```
	if len(desiredANPState.egressRules) < len(currentANPState.egressRules) || len(desiredANPState.ingressRules) < len(currentANPState.ingressRules) {
		predicateIDs := libovsdbops.NewDbObjectIDs(idType, c.controllerName,
			map[libovsdbops.ExternalIDKey]string{
				libovsdbops.ObjectNameKey: currentANPState.name,
			})
		predicateFunc := func(as *nbdb.AddressSet) bool {
			asIndex, _ := strconv.Atoi(as.ExternalIDs[libovsdbops.GressIdxKey.String()])
			return (as.ExternalIDs[libovsdbops.PolicyDirectionKey.String()] == string(libovsdbutil.ACLEgress) &&
				asIndex >= len(desiredANPState.egressRules)) ||
				(as.ExternalIDs[libovsdbops.PolicyDirectionKey.String()] == string(libovsdbutil.ACLIngress) &&
					asIndex >= len(desiredANPState.ingressRules))
		}
		asPredicate := libovsdbops.GetPredicate[*nbdb.AddressSet](predicateIDs, predicateFunc)
		ops, err = libovsdbops.DeleteAddressSetsWithPredicateOps(c.nbClient, ops, asPredicate)
		if err != nil {
			return nil, fmt.Errorf("failed to create address-set destroy ops for ANP %s, err: %v", currentANPState.name, err)
		}
	}
```
You remove that first `if` and you can do the same thing for added ANPs

Reply:
ack i see your point :) al'right i can remove that check and then directly call `constructOpsForRuleChanges` instead of `constructOpsForNewPeer` from `createNewANP` and then i remove acl cleanup for per rule from repair...
sorry for dragging this comment along so much, i should have understood what you meant (my brain needs vacation maybe..)

in that case for a real fresh create anp case, the predicate won't find any match and will just proceed with the createOrUpdate AS-es i guess. and for the update cases, where new rules are greater than older, again predicate is smart and will only add things.. ok i am sold on this idea.

I seriously don't remember why I needed that "if",, maybe because of the update case...

Reply:
The if would still be useful if you don't want to run the predicate on an update where rules were not removed, but the rest is the same.

Reply:
yes sg, updated this as per Jaime's suggestion.